sympy/sets/sets.py

```diff
@@ -278,10 +278,10 @@ def contains(self, other):
 
         """
         other = sympify(other, strict=True)
-        ret = self._contains(other)
+        ret = sympify(self._contains(other))
```

Comment:
I wouldn't blame your changes. I believe they are in line with the use of (Python and SymPy) booleans in SymPy.

It seems that the problem is caused by two unlucky choices working to counteract the effect of each other.
- The real infinities ``-oo`` and ``oo`` are considered real (and therefore also complex which is not correct). (I understand this is going to be changed after the concept of extended reals is well established.)
- An infinite interval cannot be closed at infinity (see [this](https://github.com/sympy/sympy/blob/master/sympy/sets/sets.py#L761)). (I think it should be possible to consider infinite closed intervals also on the extended real line.)

``Interval._union`` works by closing the interval at an end point contained in the union (or, more precisely, in one of its [parts](https://github.com/sympy/sympy/blob/master/sympy/sets/sets.py#L952)).

In the case at hand, ``Interval._union`` finds an open interval (``S.Reals``) whose endpoints are contained in the other set (``S.Complexes``). It tries to create a new interval that is closed (assuming that it would succeed) and returns that to the calling routine ``Union.reduce``. This finds that something has changed and repeats the call to ``Interval._union`` etc. ad infinitum.

To break the loop we should change at least one of the two choices above. It would probably be possible to allow closed infinite intervals without any disadvantage. Making the (real) infinities non-complex might lead to some regressions (though I think that should be done sooner or later).